lib/ecto/query/planner.ex

```diff
@@ -721,6 +739,23 @@ defmodule Ecto.Query.Planner do
     Map.put(query, :combinations, combinations)
   end
 
+  defp plan_ctes(%Ecto.Query{with: nil} = query, _operation, _adapter, _counter), do: query
+
+  defp plan_ctes(%Ecto.Query{with: %{queries: queries}} = query, operation, adapter, counter) do
+    queries =
+      Enum.map queries, fn
+        {name, %Ecto.Query{} = query} ->
+          {planned_query, _params, _key} = plan(query, operation, adapter, counter)
+          planned_query = planned_query |> ensure_select(true)
+          {name, planned_query}
+
+        {name, other} ->
+          {name, other}
```

Comment:
I think we need to call `plan` recursively here too so we handle any argument in the fragment.

Reply:
Didn't get it. `plan` doesn't work for `QueryExpr`.